N°6132 - Introduce API to disable tabs dynamically

dflaven opened the pull request:
_Internal PR_

dflaven committed:
N°6132 - disable tabs dynamically

Molkobain commented on js/layouts/tab-container/tab-container.js:
```diff
@@ -326,18 +330,33 @@ $(function()
+            _getTabElementFromTabIndex: function(iIdx) {
+                return this.element.children(this.js_selectors.tabs_list).children(this.js_selectors.tab_header).eq(iIdx);
+            },
+            disableTab: function(iIdx){
```
Rather pass the tab ID instead of its index (and `_getTabIndexFromTabId()`), it will improve the DX and the robustness.

Molkobain commented on js/layouts/tab-container/tab-container.js:
```diff
@@ -326,18 +330,33 @@ $(function()
+               let tabElement = this._getTabElementFromTabIndex(iIdx);
+               this._updateTabHeaderDisplay(tabElement);
+            },
+            enableTab: function(iIdx){
```
Same comment as above

Molkobain commented on js/layouts/tab-container/regular-tabs.js:
```diff
@@ -124,5 +124,15 @@ $.widget( "itop.regulartabs", $.ui.tabs, {
+		panel.css({display: 'none'});
+		this._super( index );        
+	},
+	enable: function(index) {
```
```suggestion
	// JQuery UI overload
	enable: function(index) {
```

Molkobain commented on js/layouts/tab-container/regular-tabs.js:
```diff
@@ -124,5 +124,15 @@ $.widget( "itop.regulartabs", $.ui.tabs, {
+	// JQuery UI overload
+	disable: function(index){
+		let panel = this._getPanelForTab( index );
+		panel.css({display: 'none'});
```
We rather use `.show()` / `.hide()` methods directly than CSS properties

```suggestion
		panel.hide();
```

Molkobain commented on js/layouts/tab-container/regular-tabs.js:
```diff
@@ -124,5 +124,15 @@ $.widget( "itop.regulartabs", $.ui.tabs, {
+	},
+	enable: function(index) {
+		let panel = this._getPanelForTab( index );
+		panel.css({display: 'block'});
```
```suggestion
		panel.show();
```

Molkobain commented on js/layouts/tab-container/tab-container.js:
```diff
@@ -326,18 +330,33 @@ $(function()
+                return this.element.children(this.js_selectors.tabs_list).children(this.js_selectors.tab_header).eq(iIdx);
+            },
+            disableTab: function(iIdx){
+               let tabsWidget = this.GetTabsWidget();
+               tabsWidget.disable(iIdx);
+               let tabElement = this._getTabElementFromTabIndex(iIdx);
```
We rather use `const` as the variable itself isn't supposed to be rewritten.

```suggestion
               const tabsWidget = this.GetTabsWidget();
               tabsWidget.disable(iIdx);
               const tabElement = this._getTabElementFromTabIndex(iIdx);
```

Molkobain commented on js/layouts/tab-container/scrollable-tabs.js:
```diff
@@ -377,4 +377,15 @@ $.widget( "itop.scrollabletabs", $.ui.tabs, {
+		panel.css({display: 'none'});
+		this._super( index );        
+	},
+	enable: function(index) {
```
```suggestion
	// JQuery UI overload
	enable: function(index) {
```

Molkobain commented on js/layouts/tab-container/scrollable-tabs.js:
```diff
@@ -377,4 +377,15 @@ $.widget( "itop.scrollabletabs", $.ui.tabs, {
\ No newline at end of file
+	// JQuery UI overload
+	disable: function(index){
+		let panel = this._getPanelForTab( this.tabs[index] );
+		panel.css({display: 'none'});
```
Same comments as for regular-tabs.js

Molkobain commented on js/layouts/tab-container/scrollable-tabs.js:
```diff
@@ -377,4 +377,15 @@ $.widget( "itop.scrollabletabs", $.ui.tabs, {
+		this._super( index );        
+	},
+	enable: function(index) {
+		let panel = this._getPanelForTab( this.tabs[index] );
+		panel.css({display: 'block'});
```
Same comments as for regular-tabs.js

Molkobain commented on js/layouts/tab-container/regular-tabs.js:
```diff
@@ -124,5 +124,15 @@ $.widget( "itop.regulartabs", $.ui.tabs, {
\ No newline at end of file
+	// JQuery UI overload
+	disable: function(index){
+		let panel = this._getPanelForTab( index );
```
```suggestion
		const panel = this._getPanelForTab( index );
```

Molkobain commented on js/layouts/tab-container/regular-tabs.js:
```diff
@@ -124,5 +124,15 @@ $.widget( "itop.regulartabs", $.ui.tabs, {
+		this._super( index );        
+	},
+	enable: function(index) {
+		let panel = this._getPanelForTab( index );
```
```suggestion
		const panel = this._getPanelForTab( index );
```

Molkobain commented on js/layouts/tab-container/scrollable-tabs.js:
```diff
@@ -377,4 +377,15 @@ $.widget( "itop.scrollabletabs", $.ui.tabs, {
\ No newline at end of file
+	// JQuery UI overload
+	disable: function(index){
+		let panel = this._getPanelForTab( this.tabs[index] );
```
```suggestion
		const panel = this._getPanelForTab( this.tabs[index] );
```

Molkobain commented on js/layouts/tab-container/scrollable-tabs.js:
```diff
@@ -377,4 +377,15 @@ $.widget( "itop.scrollabletabs", $.ui.tabs, {
+		this._super( index );        
+	},
+	enable: function(index) {
+		let panel = this._getPanelForTab( this.tabs[index] );
```
```suggestion
		const panel = this._getPanelForTab( this.tabs[index] );
```

Molkobain commented on js/layouts/tab-container/tab-container.js:
```diff
@@ -297,6 +297,10 @@ $(function()
             	const sTabId = oTabHeaderElem.attr('data-tab-id');
             	const oMatchingExtraTabElem = this.element.find(this.js_selectors.extra_tab_toggler+'[href="#'+sTabId+'"]');
 
+                // Disabled tabs are never added to the ExtraTabs list
+                if (oTabHeaderElem.attr('aria-disabled') == 'true') {
+                    bIsVisible = true;
+                }
```
Now that I've tried it in action, hiding the extra tab (drop down menu) when disabled feels clunking as the "normal" tab remains there. So when you resize your window or more probably when you resize the activity panel or modal the behavior is not natural.

I would rather keep it in the drop down menu with an `aria-disabled` propagated to it,
click interception and a CSS rule for the cursor.

Hipska commented:
Could someone elaborate what functionality this feature will enable?

Molkobain commented:
Enable / disable tabs of a panel in the backoffice through JS API (eg. "Impact analysis" on a "User request", or any tab of a panel for a specific use case)

Hipska commented:
With enable/disable you mean to hide it? Or just not clickable? What would be a scenario you have in mind to use this?

Molkobain commented:
> With enable/disable you mean to hide it? Or just not clickable?

The header would remain there but it would not be clickable.

> What would be a scenario you have in mind to use this?

Typically choosing an option in a tab would disable another tab.

dflaven committed:
Apply suggestions from code review

JS cleanup after review

Co-authored-by: Molkobain <lajarige.guillaume@free.fr>

dflaven committed:
Disable tabs by ID instead of index
Disabled tabs are visible (with a 'not-allowed' cursor)  instead of being hidden from the extra tabs menu.

dflaven committed:
Typo!

Molkobain commented on js/layouts/tab-container/tab-container.js:
```diff
@@ -297,17 +303,30 @@ $(function()
 
+                // Disabled tabs should be disabled in the ExtraTabs list as well
+                let bIsDisabled = false;
+                if (oTabHeaderElem.attr('aria-disabled') == 'true') {
```
```suggestion
                if (oTabHeaderElem.attr('aria-disabled') === 'true') {
```

Molkobain commented on js/layouts/tab-container/tab-container.js:
```diff
@@ -326,18 +345,35 @@ $(function()
+            _getTabElementFromTabId: function(sId) {
+                return this.element.children(this.js_selectors.tabs_list).children(this.js_selectors.tab_header+'[data-tab-id="'+sId+'"]');
+            },
+            disableTab: function(sId){
```
```suggestion
            /**
             * @param sId {string} The [data-tab-id] of the tab
             * @return {Object} The jQuery object representing the tab element
             */
            disableTab: function(sId){
```

Molkobain commented on js/layouts/tab-container/tab-container.js:
```diff
@@ -326,18 +345,35 @@ $(function()
+               const tabElement = this._getTabElementFromTabId(sId);
+               this._updateTabHeaderDisplay(tabElement); 
+            },
+            enableTab: function(sId){
```
```suggestion
            /**
             * @param sId {string} The [data-tab-id] of the tab
             * @return {Object} The jQuery object representing the tab element
             */
            enableTab: function(sId){
```

Molkobain commented on js/layouts/tab-container/tab-container.js:
```diff
@@ -326,18 +345,35 @@ $(function()
 		        return oTabElem.length === 0 ? 0 : oTabElem.prevAll().length;
-	        }
+	        },
+            _getTabElementFromTabId: function(sId) {
```
```suggestion
            /**
             * @param sId {string} The [data-tab-id] of the tab
             * @return {Object} The jQuery object representing the tab element
             *
             * @private
             */
            _getTabElementFromTabId: function(sId) {
```

Molkobain commented on js/layouts/tab-container/scrollable-tabs.js:
```diff
@@ -377,4 +377,16 @@ $.widget( "itop.scrollabletabs", $.ui.tabs, {
+	// JQuery UI overload
+	disable: function(index){
+		const panel = this._getPanelForTab( this.tabs[index] );
+		panel.css({display: 'none'});
```
```suggestion
		panel.hide();
```

Molkobain commented on js/layouts/tab-container/scrollable-tabs.js:
```diff
@@ -377,4 +377,16 @@ $.widget( "itop.scrollabletabs", $.ui.tabs, {
+	// JQuery UI overload
+	enable: function(index) {
+		const panel = this._getPanelForTab( this.tabs[index] );
+		panel.css({display: 'block'});
```
```suggestion
		panel.show();
```

Molkobain committed:
Update code to match conventions

Molkobain commented:
@dflaven I applied some code convention changes. Can you check that it still works? Then we can do the merge whenever you like (but we need to change the destination branch)

dflaven committed:
N°6132 - Fix show/hide of disabled tabs as it collided with internal methods

Molkobain commented on js/layouts/tab-container/regular-tabs.js:
```diff
@@ -127,13 +127,13 @@ $.widget( "itop.regulartabs", $.ui.tabs, {
 	disable: function(index){
 		const panel = this._getPanelForTab( index );
-		panel.hide();
+		panel.addClass('ibo-is-hidden'); // Do not use .hide() since it alters the tab state
```
```suggestion
		panel.addClass(this.css_classes.is_hidden); // Do not use .hide() since it alters the tab state
```

Molkobain commented on js/layouts/tab-container/regular-tabs.js:
```diff
@@ -127,13 +127,13 @@ $.widget( "itop.regulartabs", $.ui.tabs, {
 	enable: function(index) {
 		const panel = this._getPanelForTab( index );
-		panel.show();
+		panel.removeClass('ibo-is-hidden'); // Do not use .show() since it alters the tab state
```
```suggestion
		panel.removeClass(this.css_classes.is_hidden); // Do not use .show() since it alters the tab state
```

dflaven replied:
As far as I can see, there is no this.css_classes in the widgets scrollable-tabs and regular tabs, css classes (ui-*) are hard-coded al over the place in the code of the widget.

Molkobain replied:
Discussed IRL with Denis, the 2 widgets don't inherit from the tab-container widget and thus don't have these options.

Molkobain commented on js/layouts/tab-container/scrollable-tabs.js:
```diff
@@ -380,13 +380,13 @@ $.widget( "itop.scrollabletabs", $.ui.tabs, {
 	disable: function(index){
 		const panel = this._getPanelForTab( this.tabs[index] );
-		panel.hide();
+		panel.addClass('ibo-is-hidden'); // Do not use .hide() since it alters the tab state
```
```suggestion
		panel.addClass(this.css_classes.is_hidden); // Do not use .hide() since it alters the tab state
```

dflaven replied:
As far as I can see, there is no this.css_classes in the widgets scrollable-tabs and regular tabs, css classes (ui-*) are hard-coded al over the place in the code of the widget.

Molkobain commented on js/layouts/tab-container/scrollable-tabs.js:
```diff
@@ -380,13 +380,13 @@ $.widget( "itop.scrollabletabs", $.ui.tabs, {
 	enable: function(index) {
 		const panel = this._getPanelForTab( this.tabs[index] );
-		panel.show();
+		panel.removeClass('ibo-is-hidden'); // Do not use .show() since it alters the tab state
```
```suggestion
		panel.removeClass(this.css_classes.is_hidden); // Do not use .show() since it alters the tab state
```

dflaven replied:
As far as I can see, there is no this.css_classes in the widgets scrollable-tabs and regular tabs, css classes (ui-*) are hard-coded al over the place in the code of the widget.

Molkobain reviewed (changes requested):
Rather use `this.css_classes.is_hidden`, this allows better robustness if you change the classes name in the future.

Also beware when we merge the PR, it will be into support/3.0, not support/3.0.2
